### PR TITLE
[Refactor] Add better error handling for uploaded files

### LINKED ISSUE #4029: Better error handling for uploading files
**What problem are you trying to solve with Submitty**
Currently, many of the checks for uploading a file under course materials, and student images are not working. These checks also appear to be copied over into each function also. There are probably other places that share this copied and pasted code that should be fixed. As a result, a vague error message is sent to the user like "could not upload file(s)". 

**Describe the way you'd like to solve this problem**
Use PHP's upload errors to give a better response to the user, use CONTENT_LENGTH errors to check if a POST request was too large. Create a function (probably under fileutils) to handle validating the status of upload files.
https://www.php.net/manual/en/features.file-upload.errors.php

### PR DESCRIPTION
**Pull request by shailpatels**
This PR adds a function in FileUtils to validate uploaded files.
This properly validates filename, filesize, and any other errors that might have occurred during upload

Todo: there's still a lot of other places that are checking uploads with a copied and pasted way that's not entirely correct that need to be replaced with this function

closes #4029

### CODE REVIEW

File: site/app/libraries/FileUtils.php [hunk 1]
```diff
@@ -566,4 +566,77 @@ public static function areWordsInFile(string $file, array $words)
...
+     *
+     * @param array $files - should be in the same format as the $_FILES[] variable
+     * @return array representing the status of each file
+     * e.g. array('name' => 'foo.txt','type' => 'application/octet-stream', 'error' =>      
+     *            'success','size' => 100, 'success' => true)
+     * if $files is null returns failed => no files sent to validate
+     */
+    public static function validateUploadedFiles($files){
+
+        if(!isset($files)){
```
Comment (MasterOdin) [inline site/app/libraries/FileUtils.php]: Let's change this to be `!empty($files)` as the message applies to the empty `[]` as well.

File: site/app/libraries/FileUtils.php [hunk 2]
```diff
@@ -566,4 +566,77 @@ public static function areWordsInFile(string $file, array $words)
...
+                case UPLOAD_ERR_CANT_WRITE:
+                    $err_msg = "Failed to write file to disk";
+                    break; 
+                case UPLOAD_ERR_EXTENSION:
+                    $err_msg = "File upload stopped by extension";
+                    break; 
+                default:
+                    $err_msg = "Unknown error";
+                    $success = false;
+            }
```
Comment (MasterOdin) [inline site/app/libraries/FileUtils.php]: Can you use [ErrorMessages::uploadErrors](https://github.com/Submitty/Submitty/blob/master/site/app/libraries/ErrorMessages.php#L20) function for this switch case. I think it's just missing the `UPLOAD_ERR_INI_SIZE` and `UPLOAD_ERR_FORM_SIZE` cases.

File: site/tests/app/libraries/FileUtilsTester.php [hunk 3]
```diff
@@ -157,4 +157,204 @@ public function testGetAllDirs() {
...
+
+        $_FILES["files{$part}"]['tmp_name'][] = $tmpname;
+        $_FILES["files{$part}"]['error'][] = $err;
+
+    }
+
+    private function fillToSize($fd, $target_size){
+        fseek($fd, $target_size-1,SEEK_CUR); 
+        fwrite($fd,'a'); 
+    }
```
Comment (MasterOdin) [inline site/tests/app/libraries/FileUtilsTester.php]: Let's combine this function with the one above as they are always used together.

File: site/app/libraries/FileUtils.php (hunk 2 above)
Comment (shailpatels) [reply site/app/libraries/FileUtils.php]: Thanks for pointing that out!

**Comment by shailpatels** [comment]
@MasterOdin do you think it would be worth it to build/modify a similar function to handle validation of zip files, I didn't look to closely at how we're doing it now

File: site/app/libraries/FileUtils.php [hunk 4]
```diff
@@ -3,6 +3,7 @@
 namespace app\libraries;
 use app\exceptions\FileReadException;
 use app\libraries\Utils;
+use app\libraries\ErrorMessages;
```
Comment (MasterOdin) [inline site/app/libraries/FileUtils.php]: Remove these two lines. You don't need to add a `use ...` line for things under the same namespace as the file itself.

File: site/tests/app/libraries/FileUtilsTester.php [hunk 5]
```diff
@@ -157,4 +157,187 @@ public function testGetAllDirs() {
...
         FileUtils::recursiveRmdir($base_dir);
     }
+
+    private function buildFakeFile($fd, $filename, $part = 1, $err = 0, $target_size = 100) {
+        fseek($fd, $target_size-1,SEEK_CUR); 
+        fwrite($fd,'a'); 
+        fclose($fd);
+
+
+
```
Comment (MasterOdin) [inline site/tests/app/libraries/FileUtilsTester.php]: Shrink these three empty lines into just one empty line

**Comment by MasterOdin** [comment]
I can't really comment on it given that I'm not sure how we're handling zips. After unifying the regular file uploads, would probably be worth looking at how the zip uploading works.

**Comment by shailpatels** [comment]
> I can't really comment on it given that I'm not sure how we're handling zips. After unifying the regular file uploads, would probably be worth looking at how the zip uploading works.

Ok sounds good, I'll make the requested changes and look into zips sometime later, I assume we're doing a similar function to uploads that got copied around to different parts but I'll have to look into it more.

**Pull request merged**